scripts/backfill_source_tags_for_unregistered_contributors.py

```diff
@@ -0,0 +1,172 @@
+
+        if not dry_run:
+            set_of_user_ids_already_with_provider_source_tag = set(OSFUser.objects.filter(tags__id=source_tag.id).values_list('pk', flat=True))
+            set_of_user_ids_already_with_osf_source_tag = set(OSFUser.objects.filter(tags__id=osf_provider_source_tag.id).values_list('pk', flat=True))
+            set_bulk_create_provider_tag = set_of_user_ids_pre_npd_provider_tag.difference(set_of_user_ids_already_with_provider_source_tag)
+            set_bulk_create_osf_tag = set_of_user_ids_pre_npd_osf_tag.difference(set_of_user_ids_already_with_osf_source_tag)
+            ThroughModel.objects.bulk_create([ThroughModel(tag_id=source_tag.pk, osfuser_id=user_id) for user_id in set_bulk_create_provider_tag])
+            ThroughModel.objects.bulk_create([ThroughModel(tag_id=osf_provider_source_tag.pk, osfuser_id=user_id) for user_id in set_bulk_create_osf_tag])
+
+
+def backfill_osf_provider_tags_to_users_not_invited_but_have_no_source_tags(dry_run):
+    Tag = apps.get_model('osf', 'Tag')
+    OSFUser = apps.get_model('osf', 'OSFUser')
+    ThroughModel = OSFUser.tags.through
+    osf_provider_source_tag = Tag.all_tags.get(name='source:provider|osf', system=True)
+    source_tag_ids = Tag.all_tags.filter(name__icontains='source:', system=True).values_list('id', flat=True)
+
+    # Find not invited users with no source tags
+    id_users_with_no_source_tags = OSFUser.objects.exclude(is_invited=True).exclude(tags__id__in=source_tag_ids).values_list('pk', flat=True).order_by('pk')
+    logger.info('Number of users with no source tag {}'.format(id_users_with_no_source_tags.count()))
+
+    # Backfill OSF source tags to these users
+    logger.info('Backfilling OSF Source tags to uninvited users with no source tags')
+    paginated_users = Paginator(id_users_with_no_source_tags, 1000)
+    with tqdm(total=paginated_users.num_pages) as pbar:
+        for page_num in paginated_users.page_range:
+            through_models_to_create = []
+            for id in paginated_users.page(page_num).object_list:
+                through_models_to_create.append(ThroughModel(tag_id=osf_provider_source_tag.pk, osfuser_id=id))
+            if not dry_run:
+                ThroughModel.objects.bulk_create(through_models_to_create)
+            pbar.update()
+
+
+def main():
+    init_app(set_backends=True, routes=False)
+    dry_run = '--dry' in sys.argv
+    if not dry_run:
+        script_utils.add_file_logger(logger, __file__)
+
+    script_start_time = datetime.datetime.now()
+    logger.info('Script started time: ' + str(script_start_time))
+    backfill_source_tags_for_osf4m_unregistered_contributors(dry_run)
+    backfill_source_tags_for_nodes_and_preprints_unregistered_contributors(dry_run)
+    backfill_osf_provider_tags_to_users_not_invited_but_have_no_source_tags(dry_run)
```

Comment:
Any point in wrapping these in a transaction? it looks like you do a good job of only adding tags if they're not there already.